src/server/img_storage.rs

```diff
@@ -0,0 +1,230 @@
+                        "Bucket create error {:?}",
+                        ger
+                    ))
+                    .into());
+                }
+            }
+            Err(e) => {
+                return Err(
+                    HandlerErrorKind::Internal(format!("Bucket create error: {:?}", e)).into(),
+                )
+            }
+        };
+        // Set the permissions for the newly created bucket.
+        dbg!("Trying to grant viewing to all");
+        // grant allUsers view access
+        let all_binding = Binding {
+            role: IamRole::Standard(StandardIamRole::ObjectViewer),
+            members: vec!["allUsers".to_owned()],
+            condition: None,
+        };
+        let policy = IamPolicy {
+            bindings: vec![all_binding],
+            ..Default::default()
+        };
+        match bucket.set_iam_policy(&policy).await {
+            Ok(_) => {}
+            Err(cloud_storage::Error::Google(ger)) => {
+                if ger.errors_has_reason(&cloud_storage::Reason::Forbidden) {
+                    dbg!("Can't set permission...");
+                } else {
+                    return Err(HandlerErrorKind::Internal(format!(
+                        "Could not add read policy {:?}",
+                        ger
+                    ))
+                    .into());
+                }
+            }
+            Err(e) => {
+                return Err(HandlerErrorKind::Internal(format!(
+                    "Could not add read policy {:?}",
+                    e
+                ))
+                .into())
+            }
+        };
+        dbg!("Bucket OK");
+
+        Ok(Self {
+            // bucket: Some(bucket),
+            settings: sset,
+        })
+    }
+
+    fn gen_path(uri: &uri::Uri) -> String {
+        format!("{}{}", uri.host().expect("No host!?"), uri.path())
+    }
+
+    fn gen_public_url(&self, image_path: &str) -> String {
+        format!(
+            "{endpoint}/{project_name}/{bucket_name}/{image_path}",
+            endpoint = self.settings.endpoint,
+            project_name = self.settings.project_name,
+            bucket_name = self.settings.bucket_name,
+            image_path = image_path,
+        )
+    }
+
+    pub async fn store(&self, uri: &uri::Uri) -> HandlerResult<StoreResult> {
+        // This will absolutely fetch and store the img into the bucket.
+        // We don't do any form of check to see if it matches what we got before.
+        // If you have "Storage Legacy Bucket Writer" previous content is overwritten.
+        // (e.g. set the path to be the SHA1 of the bytes or whatever.)
+        dbg!("fetching...", &uri);
+        /*
+        // Should we preserve the name of the image?
+        let mut hasher = Sha256::new();
+        hasher.update(uri.path().as_bytes());
+        let hash = hex::encode(hasher.finalize().as_slice());
+        */
+
+        let res = reqwest::get(&uri.to_string())
```

Comment:
If/when we begin using this we'd want to reuse the same reqwest client (don't need it now)

Reply:
Yeah, I kinda view this more as example code than anything else. There are a few good optimizations I'm not doing for the sake of legibility